Integer complement in C

CBirkbeck opened the pull request:
We develop some basic API for the complement of the integers in the complex numbers.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,60 @@
+# Integer Complement
+
+We define the complement of the integers in the complex plane and give some basic lemmas about it.
+We also show that the upper half plane is mapped to the integer complement.
```
```suggestion
We also show that the upper half plane embeds into the integer complement.
```
perhaps?

b-mehta commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,60 @@
+  rw [Int.norm_eq_abs]
+  exact Int.cast_abs
+
+lemma ℂ_ℤ_Isclosed : IsClosed (((↑) : ℤ → ℂ)'' ⊤) := by
```
```suggestion
lemma ℂ_ℤ_Isclosed : IsClosed (((↑) : ℤ → ℂ) '' ⊤) := by
```

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,60 @@
+
+instance : Coe ℂ_ℤ ℂ := ⟨fun x => x.1⟩
+
+lemma upper_half_plane_ne_int (z : ℍ) : ∀ n : ℤ, z.1 ≠ n := by
```
```suggestion
lemma UpperHalfPlane.ne_int (z : ℍ) : ∀ n : ℤ, z.1 ≠ n := by
```

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,60 @@
+  have h1 := z.2
+  aesop
+
+instance coe_upp : Coe ℍ ℂ_ℤ := ⟨fun x => ⟨x, by simpa using upper_half_plane_ne_int x⟩⟩
```
```suggestion
instance : Coe ℍ ℂ_ℤ := ⟨fun x => ⟨x, by simpa using upper_half_plane_ne_int x⟩⟩
```

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,60 @@
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+def ℂ_ℤ := {z : ℂ // ¬ ∃ (n : ℤ), z = ↑n}
```
Can we give this a name and make this a (scoped) notation? Then you could use it in the names below as well

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,66 @@
+
+local notation "ℂ_ℤ " => ComplexIntegerComplement
+
+noncomputable instance : UniformSpace ℂ_ℤ  :=  instUniformSpaceSubtype
```
```suggestion
noncomputable instance : UniformSpace ℂ_ℤ := instUniformSpaceSubtype
```

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,66 @@
+
+noncomputable instance : UniformSpace ℂ_ℤ  :=  instUniformSpaceSubtype
+
+theorem Complex.closedEmbedding_coe_complex : ClosedEmbedding ((↑) : ℤ → ℂ) := by
```
Please move to `Mathlib.Analysis.Complex.Basic` as `Complex.closedEmbedding_intCast`

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,66 @@
+  rw [Int.norm_eq_abs]
+  exact Int.cast_abs
+
+lemma ℂ_ℤ_Isclosed : IsClosed (((↑) : ℤ → ℂ) '' ⊤) := by
```
Please don't use top for sets, but also it seems like this should use `Set.range` directly. Also, this should move to `Mathlib.Analysis.Complex.Basic` as `Complex.isClosed_range_intCast`

Ruben-VandeVelde commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,66 @@
+  simp only [Set.top_eq_univ, Set.image_univ]
+  exact Complex.closedEmbedding_coe_complex.isClosed_range
+
+lemma ℂ_ℤ_IsOpen : IsOpen {z : ℂ | ¬ ∃ (n : ℤ), z = ↑n} := by
```
Also move, and I think state as
```lean
lemma Complex.isOpen_compl_range_intCast : IsOpen (Set.range ((↑) : ℤ → ℂ))ᶜ :=
  Complex.isClosed_range_intCast.isOpen_compl 
```

j-loreaux reviewed (commented):
You'll need to merge master, as some things like `ClosedEmbedding` have been renamed.

CBirkbeck commented:
So I changed the definition to now make it a `Set ℂ` instead of its own type. Originally I was just following the UpperHalfPlane API, but as @loefflerd points out, this is probably not needed here. But if anyone has a good reason why it should be its own type then I'd be interested to know!

loefflerd commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,52 @@
+
+local notation "ℂ_ℤ " => IntegerComplement
+
+noncomputable instance : UniformSpace ℂ_ℤ := instUniformSpaceSubtype
```
This is not needed with the `Set` version

loefflerd commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,52 @@
+
+noncomputable instance : UniformSpace ℂ_ℤ := instUniformSpaceSubtype
+
+instance : LocallyCompactSpace ℂ_ℤ := by
+  apply IsOpen.locallyCompactSpace
+  convert Complex.isOpen_compl_range_intCast
+  rw [IntegerComplement]
+  aesop
```
Nor this – better to have a lemma `isOpen_integerComplement`

loefflerd commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,52 @@
+  rw [IntegerComplement]
+  aesop
+
+/--The coercion from the upper half plane into the IntegerComplement. -/
+@[coe] def ucoe : ℍ → ℂ_ℤ := fun z => ⟨z, by simpa using UpperHalfPlane.ne_int z⟩
+
+instance : Coe ℍ ℂ_ℤ := ⟨ucoe⟩
```
Keeping track of coercions is already a serious headache in the modular forms code; let's not add even more of them. Why not just `lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : \u z \in ℂ_ℤ`?

loefflerd commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,52 @@
+
+instance : Coe ℍ ℂ_ℤ := ⟨ucoe⟩
+
+lemma IntegerComplement_add_ne_zero (x : ℂ_ℤ) (a : ℤ) : x.1 + a ≠ 0 := by
+  intro h
+  rw [add_eq_zero_iff_eq_neg] at h
+  have := not_exists.mp x.2 (-a)
+  aesop
+
+lemma IntegerComplement_ne_zero (x : ℂ_ℤ) : x.1 ≠ 0 := by
+  simpa using IntegerComplement_add_ne_zero x 0
+
+lemma IntegerComplement_ne_one (x : ℂ_ℤ) : x.1 ≠ 1 := by
+  have := IntegerComplement_add_ne_zero x (-1 : ℤ)
+  aesop
```
I would write the hypotheses as `{x : ℂ} (hx : x \in ℂ_ℤ)` rather than using a subtype

CBirkbeck commented:
Ok I've added the suggestions from the reviews and I also changed the definition so actually be `(Set.range ((↑) : ℤ → ℂ))ᶜ` since that is what the instances are written as. I added some basic lemmas to relate to the previous definition. But I'm happy to change it back if people think this isn't ideal.

CBirkbeck commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,57 @@
+
+local notation "ℂ_ℤ " => IntegerComplement
+
+instance : IsOpen ℂ_ℤ := isOpen_compl_range_intCast
```
No it wasnt supposed to be there. I was just testing if apply? would find the result, but forgot to delete it!

urkud commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,55 @@
+open UpperHalfPlane
+
+/--The complement of the integers in `ℂ`. -/
+abbrev Complex.IntegerComplement := (Set.range ((↑) : ℤ → ℂ))ᶜ
```
This is a set, so it should use `lowerCamelCase`.
```suggestion
abbrev Complex.integerComplement := (Set.range ((↑) : ℤ → ℂ))ᶜ
```
BTW, why do you want to have it reducible?

j-loreaux replied:
It should find it if you `rw [Complex.integerComplement]` and then use `apply?`. I suspect you don't want this to be an `abbrev`, but I could be wrong.

CBirkbeck replied:
Ah yes ok that works. I changed it back to a `def`. But out of curiosity, why do you think we don't want it to be an abbrev? (I dont have much of a feel for these things).

j-loreaux replied:
I'm not totally sure that it should be `def` over `abbrev`; it's just my intuition here. I don't think there's anything strong pointing in either direction.

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,55 @@
+
+local notation "ℂ_ℤ " => IntegerComplement
+
+lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), z = ↑n} := by
+  rw [IntegerComplement]
+  aesop
```
If you turn the equality around (which you *should* do here and elsewhere) this is just `rfl`:
```suggestion
lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
```

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,55 @@
+  rw [IntegerComplement]
+  aesop
+
+lemma IntegerComplement_not_exist {x : ℂ} (hx : x ∈ ℂ_ℤ) : ¬ ∃ (n : ℤ), x = ↑n := by
+  rw [IntegerComplement_eq] at hx
+  exact hx
+
+lemma IntegerComplement_mk {x : ℂ} (hx : ¬ ∃ (n : ℤ), x = ↑n) : x ∈ ℂ_ℤ := by
+  rw [IntegerComplement_eq]
+  exact hx
```
Likewise:
```suggestion
lemma IntegerComplement_not_exist {x : ℂ} (hx : x ∈ ℂ_ℤ) : ¬ ∃ (n : ℤ), n = x := hx

lemma IntegerComplement_mk {x : ℂ} (hx : ¬ ∃ (n : ℤ), n = x) : x ∈ ℂ_ℤ := hx
```
But actually, the latter is misnamed now that it's not a subtype. I would probably just turn this into a `rw` lemma like this:
```
lemma integerComplement.mem_iff {x : ℂ} : x ∈ ℂ_ℤ ↔ ¬ ∃ (n : ℤ), n = x := Iff.rfl
```
and then you can add `alias`es if you prefer.

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,55 @@
+ rw [IntegerComplement_eq]
+ simpa using UpperHalfPlane.ne_int z
+
+lemma IntegerComplement_add_ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) (a : ℤ) : x + (a : ℂ)  ≠ 0 := by
```
You can have a stronger result here:
```suggestion
lemma integerComplement.add_coe_int_mem {x : ℂ} (a : ℤ) : x + (a : ℂ) ∈ ℂ_ℤ ↔ x ∈ ℂ_ℤ := by
```
And then your original result follows easily from this and the next result.

CBirkbeck replied:
Sure! although my proof of this is not very pretty, maybe you have some ideas how to golf it? (also is there a contrapose for iffs??)

j-loreaux replied:
Done in another suggestion.

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,55 @@
+  have := not_exists.mp hx (-a)
+  aesop
+
+lemma IntegerComplement_ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) : x ≠ 0 := by
+  simpa using IntegerComplement_add_ne_zero hx 0
```
I would namespace this, and the new proof avoids the previous lemma.
```suggestion
lemma integerComplement.ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) : x ≠ 0 :=
  fun hx' ↦ hx ⟨0, by exact_mod_cast hx'.symm⟩
```

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,55 @@
+lemma IntegerComplement_ne_zero {x : ℂ} (hx : x ∈ ℂ_ℤ) : x ≠ 0 := by
+  simpa using IntegerComplement_add_ne_zero hx 0
+
+lemma IntegerComplement_ne_one {x : ℂ} (hx : x ∈ ℂ_ℤ): x ≠ 1 := by
+  have := IntegerComplement_add_ne_zero hx (-1 : ℤ)
+  aesop
```
likewise:
```suggestion
lemma integerComplement.ne_one {x : ℂ} (hx : x ∈ ℂ_ℤ): x ≠ 1 :=
  fun hx' ↦ hx ⟨1, by exact_mod_cast hx'.symm⟩
```

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,57 @@
+
+lemma integerComplement.mem_iff {x : ℂ} : x ∈ ℂ_ℤ ↔ ¬ ∃ (n : ℤ), n = x := Iff.rfl
+
+lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : ↑z ∈ ℂ_ℤ := by
+ rw [integerComplement.mem_iff]
+ simp only [not_exists]
+ exact fun x a ↦ (UpperHalfPlane.ne_int z)  x (id (Eq.symm a))
```
```suggestion
lemma UpperHalfPlane.coe_mem_integerComplement (z : ℍ) : ↑z ∈ ℂ_ℤ :=
  not_exists.mpr fun x hx ↦ ne_int z x hx.symm 
```

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,57 @@
+ exact fun x a ↦ (UpperHalfPlane.ne_int z)  x (id (Eq.symm a))
+
+lemma integerComplement.add_coe_int_mem {x : ℂ} (a : ℤ) : x + (a : ℂ) ∈ ℂ_ℤ ↔ x ∈ ℂ_ℤ := by
+  constructor
+  · contrapose
+    simp only [Set.mem_compl_iff, Set.mem_range, not_exists, not_forall, Decidable.not_not,
+      forall_exists_index]
+    intro b hb
+    refine ⟨b + a, by simp only [Int.cast_add, hb]⟩
+  · contrapose
+    simp only [Set.mem_compl_iff, Set.mem_range, not_exists, not_forall, Decidable.not_not,
+      forall_exists_index]
+    intro b hb
+    refine ⟨b - a, by simp only [Int.cast_sub, hb, add_sub_cancel_right]⟩
```
```suggestion
  simp only [mem_iff, not_iff_not]
  exact ⟨(Exists.elim · fun n hn ↦ ⟨n - a, by simp [hn]⟩),
    (Exists.elim · fun n hn ↦ ⟨n + a, by simp [hn]⟩)⟩
```
or if you prefer it not so heavily golfed:
```suggestion
  simp only [mem_iff, not_iff_not]
  constructor
  · rintro ⟨n, hn⟩
    exact ⟨n - a, by simp [hn]⟩
  · rintro ⟨n, rfl⟩
    exact ⟨n + a, by simp⟩
```

j-loreaux commented on Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:
```diff
@@ -163,6 +163,16 @@ theorem normSq_ne_zero (z : ℍ) : Complex.normSq (z : ℂ) ≠ 0 :=
+  have h1 := z.2
+  aesop
+
+lemma ne_int (z : ℍ) : ∀ n : ℤ, z.1 ≠ n := by
+  intro n
```
The proof of `UpperHalfPlane.coe_mem_integerComplement` mildly suggests that these lemmas should be turned around, but I'm not going to push for this.

CBirkbeck replied:
Hmm sure, I'll see if it becomes annoying later on with the use case I have for this, if so I'll come back and change them.

j-loreaux reviewed (commented):
Thanks! I've supplied some alternate proofs for you.

j-loreaux commented on Mathlib/Analysis/Complex/IntegerCompl.lean:
```diff
@@ -0,0 +1,47 @@
+
+local notation "ℂ_ℤ " => integerComplement
+
+lemma IntegerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
```
```suggestion
lemma integerComplement_eq : ℂ_ℤ = {z : ℂ | ¬ ∃ (n : ℤ), n = z} := rfl
```

j-loreaux reviewed (commented):
bors d+

CBirkbeck commented:
bors r+
